lib/ExecutionEngine/Context.cpp

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/ExecutionEngine/Context.h"
+
+using namespace glow;
+
+Tensor *Context::get(Placeholder *P) {
+  auto it = map_.find(P);
+  if (it == map_.end()) {
+    return nullptr;
+  }
+
+  return it->second;
+}
+
+void Context::insert(Placeholder *P, Tensor &&T) {
+  assert(!map_.count(P) && "Placeholder already registered");
+  // Take ownership over the tensor.
+  map_[P] = new Tensor(std::move(T));
```

Comment:
T is already an rvalue reference here...  So std::move() is redundant.

Reply:
I am not sure why, but for some reason the code requires std::move. As you said, the parameter is properly marked as rvalue and I would expect things to work. However, when I remove the std::move I am getting the error below. I am not sure what's going on here. 

/Glow/lib/ExecutionEngine/Context.cpp:36:17: error: call to deleted constructor of 'glow::Tensor'
  map_[P] = new Tensor((T));
                ^      ~~~
/Glow/include/glow/Base/Tensor.h:152:3: note: 'Tensor' has been explicitly marked deleted here
  Tensor(const Tensor &other) = delete;
  ^
1 error generated.

Reply:
Oh, huh.  Interesting.  Okay.

Reply:
Ah, your original code is correct.  I was wrong to comment.  :-)

(Gleaned from the internet...) T has type rvalue reference but is an lvalue.  Therefore std::move() is necessary.

Reply:
Ah, this is complicated.
🤒 I will never fully understand this. 🤕 
Thank you for explaining this to me.